### PR TITLE
Adds stricter enforcement of Table element validation.

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This enforces all ElementSegment rules prior to creating a new module
engine. This allows flexibility and less errors inside that function.

This also uses interface as table element to pass tests with -race

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
@mathetake I picked in your https://github.com/tetratelabs/wazero/pull/353 as I couldn't find an easier way besides a polymorphic type to hold the tables at the moment. The problems with refactor it are about sharing and using the right table context in a function call. Right now, it seems like punting to interface is the best way forward.

Before I work more on tests, can you give a quick look that the approach is sound?

**Comment by mathetake** [comment]
looks fine to me

**Comment by codefromthecrypt** [comment]
ok next steps are basically backfilling tests of which there are many!

**Comment by mathetake** [comment]
https://github.com/tetratelabs/wazero/runs/5536658163?check_suite_focus=true

oh something weird  happens.. maybe double free of compiledFunctions..

**Comment by codefromthecrypt** [comment]
ok rebased... now will go back and backfill the unit tests.

File: internal/leb128/leb128.go [hunk 1]
```diff
@@ -32,7 +32,14 @@ var encodeCache = [0x80][]byte{
...
 // See https://en.wikipedia.org/wiki/LEB128#Encode_unsigned_integer
-func EncodeUint32(value uint32) (buf []byte) {
+func EncodeUint32(value uint32) []byte {
+	return EncodeUint64(uint64(value))
+}
+
+// EncodeUint64 encodes the value into a buffer in LEB128 format
+//
+// See https://en.wikipedia.org/wiki/LEB128#Encode_unsigned_integer
+func EncodeUint64(value uint64) (buf []byte) {
```
Comment (mathetake) [inline internal/leb128/leb128.go:42]: maybe we could just rename this to EncodeVarUint as the simple doesn't rely on the bit length!

Comment (mathetake) [inline internal/leb128/leb128.go:42]: maybe we could just rename this to EncodeVarUint as the simple doesn't rely on the bit length!

Comment (mathetake) [reply internal/leb128/leb128.go:42]: ah nvm!

File: internal/wasm/jit/engine.go [hunk 2]
```diff
@@ -368,14 +368,12 @@ func (e *engine) NewModuleEngine(name string, importedFunctions, moduleFunctions
 		e.addCompiledFunction(f, compiled)
 	}
 
+	for elemIdx, funcidx := range tableInit { // Initialize any elements with compiled functions
```
Comment (codefromthecrypt) [inline internal/wasm/jit/engine.go:371]: oops I forgot tests for this, will update the PR

File: internal/wasm/table.go [hunk 3]
```diff
@@ -18,68 +24,197 @@ type ElementSegment struct {
...
+//
+// Note: The global imported at globalIdx may have an offset value that is out-of-bounds for the corresponding table.
+type validatedElementSegment struct {
+	// opcode is OpcodeGlobalGet or OpcodeI32Const
+	opcode Opcode
+
+	// arg0 is the arg to opcode, which results in the offset to add to init indices.
+	//  * OpcodeGlobalGet: position in the global index namespace of an imported Global ValueTypeI32 holding the offset.
+	//  * OpcodeI32Const: a constant ValueTypeI32 offset.
+	arg0 uint32
```
Comment (mathetake) [inline internal/wasm/table.go:51]: I would drop `0` as in anyway there would be at most one arg

**Review by mathetake — APPROVED** [review]

File: internal/wasm/table.go (hunk 3 above)
Comment (codefromthecrypt) [reply internal/wasm/table.go]: sgtm!

**Comment by codefromthecrypt** [comment]
Thanks for the help @mathetake!

**Pull request merged**